mlir-sycl/include/mlir/Dialect/SYCL/Analysis/MemoryAccessAnalysis.h

```diff
@@ -0,0 +1,421 @@
+  /// Construct an offset vector with the specified number of rows.
+  OffsetVector(unsigned nRows);
+
+  /// Construct an offset vector from an initialization list.
+  OffsetVector(std::initializer_list<Value> initList);
+
+  /// Access the offset at the specified \p row.
+  Value &at(unsigned row) {
+    assert(row < nRows && "Row outside of range");
+    return offsets[row];
+  }
+
+  Value at(unsigned row) const {
+    assert(row < nRows && "Row outside of range");
+    return offsets[row];
+  }
+
+  Value &operator()(unsigned row) { return at(row); }
+  Value operator()(unsigned row) const { return at(row); }
+
+  /// Swap \p row with \p otherRow.
+  void swapRows(unsigned row, unsigned otherRow);
+
+  unsigned getNumRows() const { return nRows; }
+
+  ArrayRef<Value> getOffsets() const { return offsets; }
+
+  /// Get the offset value at the specified \p row.
+  Value getOffset(unsigned row) const;
+
+  /// Set the specified \p row to the given \p offset value.
+  void setOffset(unsigned row, Value offset);
+
+  /// Fill the offset vector with the given value \p val.
+  void fill(Value val);
+
+  /// Add an extra element at the bottom of the offset vector and set it to the
+  /// given \p offset value.
+  unsigned append(Value offset);
+
+  //===----------------------------------------------------------------------===//
+  // Queries
+  //===----------------------------------------------------------------------===//
+
+  /// Returns true if the vector contains all zeros.
+  bool isZero(DataFlowSolver &solver) const;
+
+  /// Returns true if the vector contains all zeros and the last element has
+  /// a strictly positive constant value.
+  bool isZeroWithLastElementStrictlyPositive(DataFlowSolver &solver) const;
+
+  /// Returns true if the vector contains all zeros and the last element has
+  /// value equal to the given constant \p k.
+  bool isZeroWithLastElementEqualTo(int k, DataFlowSolver &solver) const;
+
+private:
+  /// Determine whether the value at \p row is a constant integer value.
+  Optional<APInt> getConstIntegerValue(unsigned row,
+                                       DataFlowSolver &solver) const;
+
+private:
+  unsigned nRows;
+  SmallVector<Value> offsets;
+};
+
+inline raw_ostream &operator<<(raw_ostream &os, const OffsetVector &vector) {
+  llvm::interleave(
+      vector.getOffsets(), os, [&os](Value elem) { os << elem; }, " ");
+  os << "\n";
+  return os;
+}
+
+/// Describes an array access via an access matrix and an offset vector.
+/// For example consider the following access in a loop nest:
+///   for (i)
+///     for (j)
+///       ...= A[c1*i+c2][c3*j+c4];
+///
+/// The access is described by:
+///   |c1   0|   |i|   |c2|
+///   |      | * | | + |  |
+///   | 0  c3|   |j|   |c4|
+///
+template <typename OpTy> class MemoryAccess {
+  friend raw_ostream &operator<<(raw_ostream &, const MemoryAccess &);
+
+public:
+  MemoryAccess() = delete;
+
+  MemoryAccess(const OpTy &accessOp, MemoryAccessMatrix &&matrix,
```

Comment:
Technically, MemoryAccess should only take the accessOp, then matrix and offsets should be calculated?